rfcs/react-components/styles-best-practices.md

```diff
@@ -0,0 +1,599 @@
+# RFC: best practices for writing styles
+
+---
+
+[@layeshifter](https://github.com/layershifter)
+
+## Summary
+
+Fluent UI React v9 uses [Griffel][griffel] CSS-in-JS to style components. This document covers best practices for Fluent UI collaborators and library consumers.
+
+## Atomic CSS
+
+Atomic CSS is a constrasting approach to monolithic classes. In Atomic CSS every property-value is written as a single CSS rule.
+
+```css
+/* Monolithic classes */
+/* Can be applied only to a specific button */
+.button {
+  display: flex;
+  align-items: center;
+  margin-left: 8px;
+}
+
+/* Atomic CSS */
+/* Can be applied to any element that needs there rules */
+.display-flex {
+  display: flex;
+}
+.align-items-center {
+  align-items: center;
+}
+.margin-left-8px {
+  margin-left: 8px;
+}
+```
+
+[Learn more][griffel-atomic-css] about Atomic CSS.
+
+### Why Atomic CSS?
+
+Atomic CSS allows to archive:
+
+- solves [CSS specificity issues][youtube-css-specificity]
+- generates deterministic CSS classes for Server-Side Rendering
+- enables [Ahead-of-time compilation][griffel-aot] and [CSS extraction][griffel-css-extraction]
+- [smaller CSS stylesheets][youtube-atomic-css-scale] i.e. better bundle size at scale
+
+### Limitations
+
+Atomic CSS generates more classes than the approach with monolithic classes. Usually it's not a problem, but there are cases when performance [may degrade][griffel-recalc-performance], for example when elements have more than 100 classes. This does not happen usually as there is a limited set of CSS properties that is applied to a DOM element:
+
+```js
+import { makeStyles, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    display: 'flex',
+    alignItems: 'center',
+    ...shorthands.padding('4px'),
+  },
+});
+// ⬇️⬇️⬇️
+// produces 6 classes (1+1+4)
+```
+
+However, the situation can go out of control with nested selectors, [pseudo classes][mdn-pseudo-classes]/selectors and [At-rules][mdn-at-rules] causing "CSS rules explosion":
+
+```js
+import { makeStyles, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    ...shorthands.padding('4px'),
+    ...shorthands.margin('4px'),
+
+    ':hover': {
+      ...shorthands.padding('4px'),
+      ...shorthands.margin('4px'),
+    },
+
+    '::before': {
+      display: 'block',
+      content: "' '",
+    },
+
+    '@media (forced-colors: active)': {
+      ...shorthands.padding('4px'),
+      ...shorthands.margin('4px'),
+    },
+  },
+});
+// ⬇️⬇️⬇️
+// produces 14 classes ((4+4)+(4+4)+2+(4+4))
+```
+
+Such cases are unavoidable by design.
+
+- We recommend to use native pseudo classes over JS state i.e. `:hover`/`:focus` instead of manual handling of `mouseover`/`focus` events
+- An alternative to pseudo elements is more DOM elements 👎
+- There is [`.matchMedia` API][mdn-match-media], but it does not cover all At-rules
```

Comment:
I'm not sure what message you are trying to convey with theese 3 bullet points

Reply:
Good call, I removed them.